docs/go-storage/internal/iterator.md

```diff
@@ -0,0 +1,192 @@
+- When `NextObjectFunc` returns `IteratorDone`, `ObjectIterator` will not call `NextObjectFunc` anymore.
+- `ObjectIterator` will return an Object in `ObjectPage` everytime user calls `Next()`
+- `ObjectIterator` will return `IteratorDone` if no more objects.
+
+```go
+func (it *ObjectIterator) Next() (object *Object, err error) {
+   // Consume Data via index.
+   if it.index < len(it.o.Data) {
+      it.index++
+      return it.o.Data[it.index-1], nil
+   }
+   // Return IterateDone if iterator is already done.
+   if it.done {
+      return nil, IterateDone
+   }
+
+   // Reset buf before call next.
+   it.o.Data = it.o.Data[:0]
+
+   err = it.next(it.ctx, &it.o)
+   if err != nil && !errors.Is(err, IterateDone) {
+      return nil, fmt.Errorf("iterator next failed: %w", err)
+   }
+   // Make iterator to done so that we will not fetch from upstream anymore.
+   if err != nil {
+      it.done = true
+   }
+   // Return IterateDone directly if we don't have any data.
+   if len(it.o.Data) == 0 {
+      return nil, IterateDone
+   }
+   // Return the first object.
+   it.index = 1
+   return it.o.Data[0], nil
+}
+```
+
+## How to implement NextObjectFunc?
+
+Implement `NextObjectFunc` is the most complex thing in implement `Storager.List`. We will provide an example and explain why we should do this.
+
+```go
+type objectPageStatus struct {
+   delimiter string
+   maxKeys   int64
+   prefix    string
+}
+
+func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
+   input := &objectPageStatus{
+      maxKeys: 200,
+      prefix:  s.getAbsPath(path),
+   }
+   
+   var nextFn NextObjectFunc
+
+   switch {
+   case opt.ListMode.IsDir():
+      input.delimiter = "/"
+      nextFn = s.nextObjectPageByDir
+   default:
+      return nil, services.ListModeInvalidError{Actual: opt.ListMode}
+   }
+
+   return NewObjectIterator(ctx, nextFn, input), nil
+}
+
+func (s *Storage) nextObjectPageByDir(ctx context.Context, page *ObjectPage) error {
+   input := page.Status.(*objectPageStatus)
+
+   // construct API input via objectPageStatus.
+   ...
+
+   // Send API call only once.
+   output, err := s.service.ListObjectsV2WithContext(ctx, listInput)
+   if err != nil {
+      return err
+   }
+
+   // Handle and parse output to object
+   ...
+
+   // Return IterateDone when this is no more data.
+   if !aws.BoolValue(output.IsTruncated) {
+      return IterateDone
+   }
+
+   input.continuationToken = *output.NextContinuationToken
+   return nil
+}
+```
+
+- We can use `objectPageStatus` to store the status that is used to send API calls.
+- `List` has different `ListMode`, we need to support `ListModeDir` as least.
+- In `nextObjectPageByDir`:
+  - Don't use the `for` loop, our iterator already handles it.
+  - Don't read the whole list, this will consume too much memory.
+  - Don't store returned slices in `objectPageStatus`, handle and parse them directly to `page.Data`.
```

Comment:
Similarly for counting the number of files we don't need to, because we have iterator for next.
`objectPageStatus`only needs to store information about directories and files, the rest is optional